# Тема “Обучение с учителем”

###### Задание 1


Импортируйте библиотеки pandas и numpy.

In [1]:
import numpy as np
import pandas as pd
import sklearn

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.


In [2]:
from sklearn.datasets import load_boston
bostan = load_boston()
data = bostan.data
target = bostan.target
feature_names = bostan.feature_names
X = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=['price'] )

In [3]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [4]:
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.


In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.


In [6]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.


In [7]:
lr.fit(X_train, y_train)

LinearRegression()

In [8]:
y_pred = lr.predict(X_test)

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [10]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.711226005748491

###### Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.

In [12]:
from sklearn.ensemble import RandomForestRegressor

Сделайте агрумент n_estimators равным 1000,max_depth должен быть равен 12 и random_state сделайте равным 42.

In [14]:
RFRmodel = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.


In [15]:
RFRmodel.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [17]:
y_pred = RFRmodel.predict(X_test)


Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.


In [18]:
r2_score(y_test, y_pred)

0.87472606157312

Напишите в комментариях к коду, какая модель в данном случае работает лучше.


Алгоритм ML "Случайный лес"(0.87472606157312) показала наилучший результат предсказания чем линейная регрессия (0.711226005748491). Так же  насколько я понял алгритм Случайный лес намного затратнее операция чем ленейнная регресиия.

###### *Задание 3

Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.


In [19]:
?RandomForestRegressor.feature_importances_

С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.


In [20]:
sum(RFRmodel.feature_importances_)

1.0

In [24]:
pd.DataFrame(data = RFRmodel.feature_importances_.reshape(1,13), columns=feature_names)

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
0  0.031676  0.001543  0.007138  0.001236  0.014269  0.402682  0.014299   

        DIS       RAD       TAX   PTRATIO         B     LSTAT  
0  0.063973  0.005281  0.011525  0.018081  0.012451  0.415847